eth2near/utilities/src/eth_proof_generator.rs

```diff
@@ -0,0 +1,178 @@
+    }
+
+    Ok(trie)
+}
+
+fn encode_receipt(receipt: &TransactionReceipt) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+
+    if receipt.transaction_type != 0 {
+        stream.append(&receipt.transaction_type);
+    }
+
+    stream.begin_list(4);
+    stream
+        .append(&receipt.status)
+        .append(&receipt.cumulative_gas_used.0)
+        .append(&receipt.logs_bloom.0.to_vec());
+
+    stream.begin_list(receipt.logs.len());
+    for (_, log) in receipt.logs.iter().enumerate() {
+        stream.begin_list(3);
+        stream.append(&log.address.0.to_vec());
+
+        stream.begin_list(log.topics.len());
+        for (_, topic) in log.topics.iter().enumerate() {
+            stream.append(&topic.0.to_vec());
+        }
+
+        stream.append(&log.data.0);
+    }
+
+    stream.out().to_vec()
+}
+
+fn encode_log(log: &Log) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+    stream.begin_list(3);
+
+    stream.append(&log.address.0.to_vec());
+
+    stream.begin_list(log.topics.len());
+    for (_, topic) in log.topics.iter().enumerate() {
+        stream.append(&topic.0.to_vec());
+    }
+
+    stream.append(&log.data.0);
+
+    stream.out().to_vec()
+}
+
+fn encode_header(header: &BlockHeader) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+    stream.begin_unbounded_list();
+
+    stream
+        .append(&header.parent_hash.0.to_vec())
+        .append(&header.sha3_uncles.0.to_vec())
+        .append(&header.miner.0.to_vec())
+        .append(&header.state_root.0.to_vec())
+        .append(&header.transactions_root.0.to_vec())
+        .append(&header.receipts_root.0.to_vec())
+        .append(&header.logs_bloom.0.to_vec())
+        .append(&header.difficulty)
+        .append(&header.number.0)
+        .append(&header.gas_limit.0)
+        .append(&header.gas_used.0)
+        .append(&header.timestamp.0)
+        .append(&header.extra_data.0.to_vec())
+        .append(&header.mix_hash.0.to_vec())
+        .append(&header.nonce.0.to_vec());
+
+    if header.base_fee_per_gas.is_some() {
+        stream.append(&header.base_fee_per_gas.unwrap());
+    }
+
+    if header.withdrawals_root.is_some() {
+        stream.append(&header.withdrawals_root.clone().unwrap().0.to_vec());
+    }
+
+    if header.blob_gas_used.is_some() {
+        stream.append(&header.blob_gas_used.unwrap());
+    }
+
+    if header.excess_blob_gas.is_some() {
+        stream.append(&header.excess_blob_gas.unwrap());
+    }
+
+    if header.parent_beacon_block_root.is_some() {
+        stream.append(&header.parent_beacon_block_root.clone().unwrap().0.to_vec());
+    }
+
+    stream.finalize_unbounded_list();
+
+    stream.out().to_vec()
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use hasher::Hasher;
+
+    #[test]
+    fn generate_proof() {
+        const RPC_URL: &'static str = "https://eth.llamarpc.com";
+        const TX_HASH: &'static str = "0x9298954a9db8026ca28bce4d71ffb7ba0aac70e91f0667ffb7398c67e60b84fa";
```

Comment:
Please add more tests for transactions from different epochs(upgrades), e.g. London, Shapella, and Dencun, and compare the results with the proofs generated by the events relayer.

Reply:
> I intended to mock RPC requests and add a few tests for the client as well ass more general test cases for proof generation. Would that be the correct approach?

For now, it is okay to use the public RPC and make the requests in the tests; we can improve it later in a separate PR by storing the response in the git lfs storage.

Reply:
> Supporting older blocks
> I tested this on post merge and post-4844 blocks. Older blocks may not work (for example because of different handling of difficulty field). If needed I'll test and add support for pre-merge.

Ok, let's ignore pre-merge blocks, and just handle post-merge blocks

Reply:
Added tests in da40e8ef5a43a364b3e189e9e094293b17da35a4